src/Workspaces/Core/Desktop/Workspace/Host/TemporaryStorage/TemporaryStorageServiceFactory.cs

```diff
@@ -42,19 +93,69 @@ public ITemporaryTextStorage CreateTemporaryTextStorage(CancellationToken cancel
+                return new TemporaryTextStorage(this, storageName, offset, size, encoding);
             }
 
             public ITemporaryStreamStorage CreateTemporaryStreamStorage(CancellationToken cancellationToken)
             {
                 return new TemporaryStreamStorage(this);
             }
 
-            public ITemporaryStreamStorage AttachTemporaryStreamStorage(string storageName, long size, CancellationToken cancellationToken)
+            public ITemporaryStreamStorage AttachTemporaryStreamStorage(string storageName, long offset, long size, CancellationToken cancellationToken)
+            {
+                return new TemporaryStreamStorage(this, storageName, offset, size);
+            }
+
+            /// <summary>
+            /// Allocate shared storage of a specified size.
+            /// </summary>
+            /// <remarks>
+            /// <para>"Small" requests are fulfilled from oversized memory mapped files which support several individual
+            /// storage units. Larger requests are allocated in their own memory mapped files.</para>
+            /// </remarks>
+            /// <param name="size">The size of the shared storage block to allocate.</param>
+            /// <returns>A <see cref="MemoryMappedInfo"/> describing the allocated block.</returns>
+            private MemoryMappedInfo CreateTemporaryStorage(long size)
+            {
+                if (size >= SingleFileThreshold)
+                {
+                    // Larger blocks are allocated separately
+                    var mapName = CreateUniqueName(size);
+                    var storage = MemoryMappedFile.CreateNew(mapName, size);
+                    return new MemoryMappedInfo(new ReferenceCountedDisposable<MemoryMappedFile>(storage), mapName, 0, size);
+                }
+
+                lock (_gate)
+                {
+                    // Obtain a reference to the memory mapped file, creating one if necessary. If a reference counted
+                    // handle to a memory mapped file is obtained in this section, it must either be disposed before
+                    // returning or returned to the caller who will own it through the MemoryMappedInfo.
+                    var reference = _weakFileReference.TryAddReference();
+                    if (reference == null || _offset + size > _fileSize)
+                    {
+                        var mapName = CreateUniqueName(MultiFileBlockSize);
+                        var file = MemoryMappedFile.CreateNew(mapName, MultiFileBlockSize);
+
+                        reference = new ReferenceCountedDisposable<MemoryMappedFile>(file);
+                        _weakFileReference = new ReferenceCountedDisposable<MemoryMappedFile>.WeakReference(reference);
+                        _name = mapName;
+                        _fileSize = MultiFileBlockSize;
+                        _offset = 0;
+                    }
```

Comment:
i feel like falling through ehre makes things more difficult.  Can we just ahve two codepaths:  one for when you can reuse the existing file reference, and one for whne you need to make a fresh reference.